### PR TITLE
actions: run unit-tests and linting in GH actions

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Use GitHub actions for the static code analysis and unit tests
Bump go version

## Why is it important?

This will help with running faster builds in the CI

## Issue 

Requires https://github.com/elastic/e2e-testing/pull/3139

### CODE REVIEW

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: elastic/apm-pipeline-library/.github/actions/docker-login@gh_actions
+        with:
+          registry: docker.elastic.co
+          secret: secret/observability-team/ci/docker-registry/prod
+          url: ${{ secrets.VAULT_ADDR }}
+          roleId: ${{ secrets.VAULT_ROLE_ID }}
+          secretId: ${{ secrets.VAULT_SECRET_ID }}
```
Comment (v1v) [inline .github/workflows/pull-request.yml:53]: This won't work when running on PRs, so need to provide an alternative

File: .github/workflows/pull-request.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      - name: Fetch Go version from .go-version
+        run: echo "GO_VERSION=$(cat .go-version)" >> $GITHUB_ENV
+      - uses: actions/setup-python@v3
+      - uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GO_VERSION }}
+      - name: Install dependencies
+        run: |
+          go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.50.0
+          go install golang.org/x/lint/golint
```
Comment (v1v) [inline .github/workflows/pull-request.yml:28]: Force to bump the version to 1.18 so I could use `go install`

File: .go-version [hunk 3]
```diff
@@ -1 +1 @@
-1.17
+1.18
```
Comment (v1v) [inline .go-version:1]: I guess this should not cause any issues, but let's wait for the whole e2e-testing runs in the CI

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -52,7 +52,6 @@ repos:
             (?x)^(
                 .*/?\.*.(tar.gz)
             )$
-    -   id: check-jjbb
```
Comment (v1v) [inline .pre-commit-config.yaml:55]: Remove the dependency with the internal docker registry, as the docker image is in the private repo and vault/docker login with secrets won't work for PRs.

File: .github/workflows/pull-request.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  pull_request:
+  push:
+
+jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    env:
+      JENKINS_URL: https://beats-ci.elastic.co/
+    steps:
+      - uses: actions/checkout@v3
```
Comment (v1v) [inline .github/workflows/pull-request.yml:18]: https://github.com/elastic/apm-pipeline-library/pull/1924 could be useful in addition to https://docs.github.com/en/actions/creating-actions/creating-a-composite-action#testing-out-your-action-in-a-workflow

**Review by kuisathaverat — APPROVED** [review]

File: .github/workflows/junit-test-reporter.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+jobs:
+  report:
+    runs-on: ubuntu-latest
+    # Assign permissions for unit tests to be reported.
+    # See https://github.com/dorny/test-reporter/issues/168
+    permissions:
+      statuses: write
+      checks: write
+      contents: write
+      pull-requests: write
```
Comment (v1v) [inline .github/workflows/junit-test-reporter.yml:23]:
```suggestion
```

File: .github/workflows/junit-test-reporter.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    steps:
+      - uses: dorny/test-reporter@c9b3d0e2bd2a4e96aaf424dbaa31c46b42318226
+        with:
+          artifact: test-results            # artifact name
+          name: JUnit Tests                 # Name of the check run which will be created
+          path: "**/*.xml"                    # Path to test results (inside artifact .zip)
+          reporter: java-junit              # Format of test results
+          token: ${{ secrets.GITHUB_TOKEN }} # GitHub token
```
Comment (v1v) [inline .github/workflows/junit-test-reporter.yml:31]:
```suggestion
```

File: .github/workflows/junit-test-reporter.yml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      statuses: write
+      checks: write
+      contents: write
+      pull-requests: write
+    steps:
+      - uses: dorny/test-reporter@c9b3d0e2bd2a4e96aaf424dbaa31c46b42318226
+        with:
+          artifact: test-results            # artifact name
+          name: JUnit Tests                 # Name of the check run which will be created
+          path: "**/*.xml"                    # Path to test results (inside artifact .zip)
```
Comment (v1v) [inline .github/workflows/junit-test-reporter.yml:29]:
```suggestion
          path: "**/*.xml"                   # Path to test results (inside artifact .zip)
```

File: .github/workflows/pull-request.yml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+on:
+  pull_request:
+  push:
+
+jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    env:
+      JENKINS_URL: https://beats-ci.elastic.co/
```
Comment (cachedout) [inline .github/workflows/pull-request.yml:16]: What's this being used for?

Comment (v1v) [reply .github/workflows/pull-request.yml:16]: for linting the Jenkins pipelines, there is a pre-commit validation we provide in https://github.com/elastic/apm-pipeline-library/blob/b23be9e84d84f12f8586e73a0490a07fcfd2f221/.pre-commit-hooks.yaml#L22-L29

Eventually to be removed once it's not Jenkins based

File: .github/workflows/junit-test-reporter.yml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+jobs:
+  report:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: dorny/test-reporter@c9b3d0e2bd2a4e96aaf424dbaa31c46b42318226
+        with:
+          artifact: test-results            # artifact name
+          name: JUnit Tests                 # Name of the check run which will be created
+          path: "**/*.xml"                   # Path to test results (inside artifact .zip)
+          reporter: java-junit              # Format of test results
```
Comment (v1v) [inline .github/workflows/junit-test-reporter.yml:23]: Wait for https://github.com/elastic/apm-pipeline-library/pull/1941

**Review by cachedout — APPROVED** [review]

File: .github/workflows/junit-test-reporter.yml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+      - completed
+
+permissions:
+  contents: read
+
+jobs:
+  report:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: dorny/test-reporter@c9b3d0e2bd2a4e96aaf424dbaa31c46b42318226
```
Comment (v1v) [inline .github/workflows/junit-test-reporter.yml:18]: Use https://github.com/elastic/apm-pipeline-library/tree/main/.github/actions/test-report

```suggestion
      - uses: elastic/apm-pipeline-library/.github/actions/test-report@gh_actions
```

**Pull request merged**